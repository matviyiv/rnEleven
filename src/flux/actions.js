import moment from 'moment';
import _ from 'lodash';
import firebase from '../firebase-setup.js';
import {fetchGet} from '../utils';

export const constants = {
  SERVICES_LOADING: 'SERVICES_LOADING',
  SERVICES_LOADED: 'SERVICES_LOADED',
  SELECT_SERVICE: 'SELECT_SERVICE',
  SERVICES_FAILED: 'SERVICES_FAILED',
  MASTERS_LOADING: 'MASTERS_LOADING',
  MASTERS_LOADED: 'MASTERS_LOADED',
  MASTERS_FAILED: 'MASTERS_FAILED',
  SELECT_MASTER_NEXT_DATE: 'SELECT_MASTER_NEXT_DATE',

  BOOKING_SUBMIT: 'BOOKING_SUBMIT',
  BOOKING_SUBMITED: 'BOOKING_SUBMITED',
  BOOKING_FAILED: 'BOOKING_FAILED',
  BOOKING_CLEAR: 'BOOKING_CLEAR',
  SAVE_BOOKING_USER: 'SAVE_BOOKING_USER',
  BOOKING_DELETED: 'BOOKING_DELETED',
  BOOKING_DELETED_FAILED: 'BOOKING_DELETED_FAILED',
  BOOKING_DELETE_SERVICE: 'BOOKING_DELETE_SERVICE',
  BOOKING_UPDATED: 'BOOKING_UPDATED',
  BOOKING_UPDATE_FAILED: 'BOOKING_UPDATE_FAILED',

  MASTERS_TIME_LOADING: 'MASTERS_TIME_LOADING',
  MASTERS_TIME_LOADED: 'MASTERS_TIME_LOADED',
  MASTERS_TIME_ERROR: 'MASTERS_TIME_ERROR',
  
  ALL_EVENTS_LOADING: 'ALL_EVENTS_LOADING',
  ALL_EVENTS_LOADED: 'ALL_EVENTS_LOADED',
  ALL_EVENTS_FAILED: 'ALL_EVENTS_FAILED',

  AUTH_LOADING: 'AUTH_LOADING',
  AUTH_DONE: 'AUTH_DONE',
  AUTH_FAILED: 'AUTH_FAILED',
  LOGOUT: 'LOGOUT',

  SUB_SUCCESS: 'SUB_SUCCESS',
  SUB_FAILED: 'SUB_FAILED',
  SUB_START: 'SUB_START',

  DB_BOOKING_SUBSCRIBE: 'DB_BOOKING_SUBSCRIBE',
  DB_BOOKING_UPDATE: 'DB_BOOKING_UPDATE',

  SET_LAGUAGE: 'SET_LAGUAGE',
  LOCALIZATION_LOADED: 'LOCALIZATION_LOADED',
};

export function loadServices() {
  return dispatch => {
    dispatch({
      type: constants.SERVICES_LOADING
    });
    firebase.database().ref('lviv/services').once('value')
      .then((services) => {
        console.log('services loaded', services.val());
        dispatch({
          type: constants.SERVICES_LOADED,
          data: services.val()
        });
      })
      .catch((error) => {
        console.error('action loadServices failed', error);
        dispatch({
          type: constants.SERVICES_FAILED,
          error: error
        });
      });
  };
}

export function selectService(data) {
  return {
    type: constants.SELECT_SERVICE,
    data
  };
}

export function selectMasterNextDate(data) {
  return {
    type: constants.SELECT_MASTER_NEXT_DATE,
    data
  };
}

export function loadMasters() {
  return dispatch => {
    dispatch({
      type: constants.MASTERS_LOADING
    });
    firebase.database()
      .ref('lviv/masters')
      .once('value')
      .then((masters) => {
        const mastersList = masters.val();
        console.log('masters loaded', mastersList);
        dispatch({
          type: constants.MASTERS_LOADED,
          data: mastersList
        });
      })
      .catch((error) => {
        console.error('action loadMasters failed', error);
        dispatch({
          type: constants.MASTERS_FAILED,
          error: error
        });
      });
  };
}

export function getMastersTime(mastersList, _date_) {
  return dispatch => {
    dispatch({
      type: constants.MASTERS_TIME_LOADING
    });
    const date = moment(_date_);
    if (mastersList) {
      return Promise.all(
        mastersList.map((master) => firebase.database()
          .ref(`lviv/mastersTime/${master.id}/${date.get('year')}/${date.get('month')}/${date.get('date')}`).once('value'))
      )
      .then((timeList) => timeList.map((time) => time.val()))
      .then((result) => {
        dispatch({
          type: constants.MASTERS_TIME_LOADED,
          data: {mastersList, result, date}
        });
      })
      .catch((e) => {
        dispatch({
          type: constants.MASTERS_TIME_ERROR,
          error: e
        });
      });
    }
  };
}

export function getAllEvents() {
  return dispatch => {
    const date = moment().subtract(1, 'month');
    dispatch({
      type: constants.ALL_EVENTS_LOADING
    });

    Promise.all([
      loadMastersRequest(),
      loadBookingsRequest(date),
    ])
    .then(([masters, bookings]) => {
      dispatch({
        type: constants.MASTERS_LOADED,
        data: masters.val()
      });
      dispatch({
        type: constants.ALL_EVENTS_LOADED,
        data: bookings.val()
      });
    })
    .catch((error) => {
      console.error('action getAllEvents failed', error);
      dispatch({
        type: constants.ALL_EVENTS_FAILED,
        error: error
      });
    });
  };
}

export function subscribeForNewBookings() {
  return dispatch => {
    dispatch({
      type: constants.DB_BOOKING_SUBSCRIBE
    });
    firebase.database()
      .ref('lviv/bookings')
      .on('child_added', function(snapshot) {
        dispatch({
          type: constants.DB_BOOKING_UPDATE,
          data: {
            booking: snapshot.val(),
            bookingId: snapshot.key
          }
        });
      });
  }
}

function loadMastersRequest() {
  return firebase.database()
    .ref('lviv/masters')
    .once('value');
}

function loadBookingsRequest(date) {
  return firebase.database()
    .ref('lviv/bookings')
    .orderByChild('timestamp')
    .startAt(date.toDate().getTime())
    .once('value');
}

export function submitBooking(_booking_) {
  return dispatch => {
    const booking = {..._booking_};
    booking.timestamp = moment().toDate().getTime();
    booking.status = 'new';
    dispatch({
      type: constants.BOOKING_SUBMIT,
      data: {booking}
    });
    const mastersData = getMasterTime(booking.selectedServices);
    Promise.all([
      firebase.database().ref('lviv/mastersTime')
        .update(mastersData),
      firebase.database().ref('lviv/bookings')
        .push(booking)
    ])
      .then(() => {
        console.log('submitBooking done');
        dispatch({
          type: constants.BOOKING_SUBMITED
        });
      })
      .catch((error) => {
        console.error('action submitBooking failed', error);
        dispatch({
          type: constants.BOOKING_FAILED,
          error: error
        });
      });
  };
}

export function saveBookingUser(data) {
  return {
    type: constants.SAVE_BOOKING_USER,
    data
  };
}

export function clearBooking() {
  return {
    type: constants.BOOKING_CLEAR,
  };
}

export function deleteBoking(bookingId) {
  return dispatch => {
    firebase.database()
    .ref('lviv/bookings/' + bookingId)
    .once('value')
    .then((_booking) => {
      let booking = _booking.val();
      const mastersData = getMasterTime(booking.selectedServices);

      booking.status = 'deleted';
      return Promise.all([
        _.keys(mastersData).map((key) => firebase.database().ref('lviv/mastersTime/' + key).remove()),
        firebase.database().ref('lviv/bookings/' + bookingId)
          .update(booking)
      ]);
    })
    .then(() => {
      dispatch({
        type: constants.BOOKING_DELETED,
        data: {bookingId}
      });
    })
    .catch((error) => {
      dispatch({
        type: constants.BOOKING_DELETED_FAILED,
        data: {error}
      });
    });
    
  };
}

export function updateBooking(bookingId, booking, subServiceId) {
  return dispatch => {
    firebase.database().ref('lviv/bookings/' + bookingId)
      .update(booking)
      .then(() => {
        dispatch({type: constants.BOOKING_UPDATED, data: {booking, subServiceId, bookingId}})
      })
      .catch((error) => {
        dispatch({
          type: constants.BOOKING_UPDATE_FAILED,
          error: 'Failed to update booking' + error
        });
      });
  }
}

export function authenticate(email, password) {
  return dispatch => {
    dispatch({
      type: constants.AUTH_LOADING,
    });
    firebase.auth().signInWithEmailAndPassword(email, password)
      .then(() => {
        dispatch({
          type: constants.AUTH_DONE,
          data: {email, id: firebase.auth().currentUser.uid}
        });
      })
      .catch((error) => {
        dispatch({
          type: constants.AUTH_FAILED,
          error
        });
      });
  };
}

export function logout() {
  return dispatch => {
    firebase.auth().signOut()
    .then(() => {
      dispatch({
        type: constants.LOGOUT
      });
    });
  };
}

export function subscribe(email) {
  return dispatch => {
    dispatch({
      type: constants.SUB_START
    });
    firebase.database().ref('lviv/subscriptions')
      .push({
        date: moment().toDate().getTime(),
        email
      })
      .then(() => dispatch({type: constants.SUB_SUCCESS}))
      .catch((error) => dispatch({type: constants.SUB_FAILED, data: {error}}));
  }
}

export function deleteSelectedService(serviceId) {
  return {
    type: constants.BOOKING_DELETE_SERVICE,
    data: {serviceId}
  }
}

function getMasterTime(bookings) {
  return _.reduce(bookings, (result, booking) => {
    let date = moment(booking.dateStart);
    result[`${booking.masterId}/${date.get('year')}/${date.get('month')}/${date.get('date')}/${date.get('hour')}/${date.get('minute')}`] = {name: booking.name, duration: booking.duration};
    return result;
  }, {});
}

export function changeLanguage(locale, str) {
  return dispatch => {
    if (str[locale]) {
      return dispatch({
        type: constants.SET_LAGUAGE,
        data: {locale}
      });
    }
   
    fetchGet(`/localization/${locale}.json`)
      .then((localization) => {
        dispatch({
          type: constants.LOCALIZATION_LOADED,
          data: {localization, locale}
        });
      })
  }
}
