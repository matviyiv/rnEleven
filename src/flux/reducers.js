import {constants} from './actions';
import moment from 'moment';
import _ from 'lodash';

const initialState = {
  services: {},
  masters: {},
  booking: {
    selectedServices: {}
  },
  allEvents: {
    list: [],
    loading: false,
  },
  auth: {
    status: '',
    loading: false,
    email: '',
  },
  subscription: {},
  db: {}
};
const {
  SERVICES_LOADING,
  SERVICES_LOADED,
  MASTERS_LOADED,
  MASTERS_LOADING,
  SELECT_MASTER_NEXT_DATE,
  MASTERS_TIME_LOADED,
  SAVE_BOOKING_USER,
  ALL_EVENTS_LOADED,
  ALL_EVENTS_LOADING,
  BOOKING_DELETED,
  BOOKING_SUBMITED,
  BOOKING_CLEAR,
  BOOKING_DELETE_SERVICE,
  BOOKING_UPDATED,
  AUTH_LOADING,
  AUTH_DONE,
  LOGOUT,
  SUB_START,
  SERVICES_FAILED,
  BOOKING_FAILED,
  BOOKING_DELETED_FAILED,
  MASTERS_TIME_ERROR,
  ALL_EVENTS_FAILED,
  AUTH_FAILED,
  DB_BOOKING_UPDATE,
  DB_BOOKING_SUBSCRIBE,
} = constants;

export function appReducer(state = initialState, action) {
  const actions = {
    [SERVICES_LOADING]: (st) => {
      st.services.loading = true;
      return { ...st };
    },
    [SERVICES_LOADED]: (st, data) => {
      st.services.loading = false;
      st.services.list = data;
      return { ...st };
    },
    [MASTERS_LOADING]: (st) => {
      st.masters.loading = true;
      return {...st};
    },
    [MASTERS_LOADED]: (st, data) => {
      st.masters.loading = false;
      st.masters.list = data;
      return { ...st };
    },
    [SELECT_MASTER_NEXT_DATE]: (st, {serviceId, masterId, date}) => {
      const service = findSubService(st.services.list, serviceId);
      const dateEnd = moment(date).add(service.duration, 'hours');
      const selectedService = {
        masterId: masterId,
        dateStart: moment(date).toString(),
        dateEnd: moment(dateEnd).toString(),
        name: service.name,
        en: service.en,
        serviceId: serviceId,
        duration: service.duration,
      };

      st.booking.selectedServices[serviceId + '+' + date.valueOf()] = selectedService;

      return { ...st };
    },
    [MASTERS_TIME_LOADED]: (st, {mastersList, result, date}) => {
      const time = moment(date).format('YYYY/M/D');
      mastersList.forEach((master, index) => {
        _.setWith(st.masters.list[master.id], `booked.${time}`, result[index], Object);
      });
      return { ...st };
    },
    [BOOKING_SUBMITED]: (st) => {
      st.booking = initialState.booking;
      return {...st};
    },
    [SAVE_BOOKING_USER]: (st, {name, phone, notes}) => {
      st.booking.name = name;
      st.booking.phone = phone;
      st.booking.notes = notes;
      st.booking.status = 'active';
      return {...st};
    },
    [BOOKING_CLEAR]: (st) => {
      st.booking.selectedServices = {};
      return {...st};
    },
    [ALL_EVENTS_LOADING]: (st) => {
      st.allEvents.loading = true;
      return {...st};
    },
    [ALL_EVENTS_LOADED]: (st, bookings) => {
      let eventsList = st.allEvents.list;
      _.forEach(bookings, (booking, bookingId) => {
        if (booking.status === 'deleted') {return;}
        eventsList = [].concat(eventsList, prepareCalendarEvent(booking, bookingId, st.masters));
      });
      st.allEvents.loading = false;
      st.allEvents.list = eventsList;
      return {...st};
    },
    [BOOKING_DELETED]: (st, {bookingId}) => {
      let index = _.findIndex(st.allEvents.list, {bookingId});
      delete st.allEvents.list[index];
      return {...st};
    },
    [BOOKING_UPDATED]: (st, {booking, bookingId, subServiceId}) => {
      let eventsList = st.allEvents.list;
      let index = _.findIndex(eventsList, {subServiceId: subServiceId});
      eventsList[index] = _.find(prepareCalendarEvent(booking, bookingId, st.masters), {subServiceId: subServiceId});
      st.allEvents.list = eventsList;
      return {...st};
    },
    [AUTH_LOADING]: (st) => {
      st.auth.loading = true;
      return {...st};
    },
    [AUTH_DONE]: (st, {email, id}) => {
      st.auth.email = email;
      st.auth.id = id;
      st.auth.status = 'success';
      st.auth.loading = false;
      return {...st};
    },
    [LOGOUT]: (st) => {
      st.auth.email = '';
      st.auth.status = '';
      return {...st};
    },
    [SUB_START]: (st) => {
      st.subscription.success = true;
      return {...st};
    },
    [BOOKING_DELETE_SERVICE]: (st, {serviceId}) => {
      const booking = {...st.booking};
      delete booking.selectedServices[serviceId];
      st.booking = booking;
      return {...st};
    },
    [DB_BOOKING_SUBSCRIBE]: (st, data) => {
      st.db.subscribed = true;
      return {...st};
    },
    [DB_BOOKING_UPDATE]: (st, {booking, bookingId}) => {
      if (booking.status === 'deleted' || !st.masters.list) {return st;}
      let eventsList = st.allEvents.list;
      if (_.find(eventsList, {bookingId: bookingId})) {return st;}
      eventsList = [].concat(eventsList, prepareCalendarEvent(booking, bookingId, st.masters));
      st.allEvents.loading = false;
      st.allEvents.list = eventsList;
      showNotification(booking);
      return {...st};
    },
    default: (st) => st
  },
  modifier = actions[action.type] || actions.default;
  if ([
      SERVICES_FAILED,
      BOOKING_FAILED,
      BOOKING_DELETED_FAILED,
      MASTERS_TIME_ERROR,
      ALL_EVENTS_FAILED,
      AUTH_FAILED,
    ].indexOf(action.type) > -1) {console.error(action.error || action.data.error);}
  const newState = modifier(state, action.data);
  console.log('----', action.type, newState);
  return newState;
}

function findSubService(services, subServiceId) {
  for (let serviceIndex in services) {
    let subservice = services[serviceIndex].sub.find((sub) => sub.id === subServiceId);
    if (subservice) {return subservice;}
  }
}

function prepareCalendarEvent(booking, bookingId, masters) {
  return _.map(booking.selectedServices, (service, subServiceId) => ({
    title: `Сервіс: ${service.name} майстер: ${_.get(masters, `list[${service.masterId}].name`, 'removed')} клієнт: ${booking.name} тел: ${booking.phone} дод: ${booking.notes}`,
    start: new Date(service.dateStart),
    end: new Date(service.dateEnd),
    desc: `${booking.name} ${booking.phone}`,
    masterId: service.masterId,
    bookingId,
    booking,
    subServiceId,
  }));
}

function showNotification(booking) {
  new Notification('Нове бронювання', {
    body: JSON.stringify(booking.selectedServices)
  })
}