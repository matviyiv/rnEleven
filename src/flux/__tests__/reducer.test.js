import moment from 'moment';
import _ from 'lodash';

import * as reducers from '../reducers';
import * as actions from '../actions';

describe('reducers', () => {
  let st;

  beforeEach(() => {
    st = {
      booking: {
        selectedServices: {}
      },
      services: {
        id: 's2',
        list: [{
          sub: [{
            id: 's2_1',
            duration: 1,
            name: 'service2'
          }]
        }]
      }
    }
  })
  it('SELECT_MASTER_NEXT_DATE', () => {
    const data = {
      serviceId: 's2_1',
      masterId: 'e1',
      date: moment(new Date('Wed Jun 14 2017 11:00:00 GMT+0300 (EEST)')).hours(11).minutes(0).seconds(0)
    };
    const result = reducers.appReducer(st, actions.selectMasterNextDate(data));
    expect(result.booking.selectedServices).toMatchSnapshot();
  });
});