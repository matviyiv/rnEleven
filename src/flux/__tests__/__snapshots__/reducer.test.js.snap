// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers SELECT_MASTER_NEXT_DATE 1`] = `
Object {
  "s2_1+1497427200000": Object {
    "dateEnd": "Wed Jun 14 2017 12:00:00 GMT+0300",
    "dateStart": "Wed Jun 14 2017 11:00:00 GMT+0300",
    "duration": 1,
    "masterId": "e1",
    "name": "service2",
  },
}
`;
